enterprise/internal/batches/service/service.go

```diff
@@ -1472,3 +1472,95 @@ func (s *Service) RetryBatchSpecExecution(ctx context.Context, opts RetryBatchSp
 
 	return nil
 }
+
+type GetAvailableBulkOperationsOpts struct {
+	BatchChange int64
+	Changesets  []int64
+}
+
+// GetAvailableBulkOperations returns all bulk operations that can be carried out
+// on an array of changesets.
+func (s *Service) GetAvailableBulkOperations(ctx context.Context, opts GetAvailableBulkOperationsOpts) ([]string, error) {
+	bulkOperationsCounter := map[btypes.ChangesetJobType]int{
+		btypes.ChangesetJobTypeClose:     0,
+		btypes.ChangesetJobTypeComment:   0,
+		btypes.ChangesetJobTypeDetach:    0,
+		btypes.ChangesetJobTypeMerge:     0,
+		btypes.ChangesetJobTypePublish:   0,
+		btypes.ChangesetJobTypeReenqueue: 0,
+	}
+
+	changesets, _, err := s.store.ListChangesets(ctx, store.ListChangesetsOpts{
+		IDs:          opts.Changesets,
+		EnforceAuthz: true,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, changeset := range changesets {
+		isChangesetArchived := changeset.ArchivedIn(opts.BatchChange)
+		isChangesetDraft := changeset.ExternalState == btypes.ChangesetExternalStateDraft
+		isChangesetOpen := changeset.ExternalState == btypes.ChangesetExternalStateOpen
+		isChangesetClosed := changeset.ExternalState == btypes.ChangesetExternalStateClosed
+		isChangesetMerged := changeset.ExternalState == btypes.ChangesetExternalStateMerged
+		isChangesetPublished := changeset.PublicationState == btypes.ChangesetPublicationStateUnpublished
+
+		// can changeset be published
+		isChangesetPublishable := isChangesetPublished || isChangesetDraft || isChangesetOpen
+		isChangesetCommentable := isChangesetOpen || isChangesetDraft || isChangesetMerged || isChangesetClosed
+		isChangesetClosable := isChangesetOpen || isChangesetDraft || isChangesetMerged
+
+		// check what operations this changeset support, most likely from the state
+		// so get the changeset then derive the operations from it's state
+
+		// DETACH
+		if isChangesetArchived {
+			bulkOperationsCounter[btypes.ChangesetJobTypeDetach] += 1
+		}
+
+		// REENQUEUE
+		if !isChangesetArchived && changeset.ReconcilerState == btypes.ReconcilerStateFailed {
+			bulkOperationsCounter[btypes.ChangesetJobTypeReenqueue] += 1
+			continue
+		}
+
+		// PUBLISH
+		if !isChangesetArchived && isChangesetPublishable {
```

Comment:
I think this matches it better

```suggestion
		if !isChangesetArchived {
```

Reply:
Oh make sense. I'll update.